### PR TITLE
percpu reg store this_task

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
We can utilize percpu storage to hold information about the             
current running task. If we intend to implement this feature, we would  
need to define two macros that help us manage this percpu information   
effectively.

up_this_task: This macro is designed to read the contents of the percpu 
              register to retrieve information about the current        
              running task.This allows us to quickly access             
              task-specific data without having to disable interrupts,  
              access global variables and obtain the current cpu index. 

up_update_task: This macro is responsible for updating the contents of  
                the percpu register.It is typically called during       
                initialization or when a context switch occurs to ensure
                that the percpu register reflects the information of the
                newly running task.
## Impact
arm64

## Testing
    Configuring NuttX and compile:
    $ ./tools/configure.sh -l qemu-armv8a:nsh_smp
    $ make
    Running with qemu
    $ qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic \
       -machine virt,virtualization=on,gic-version=3 \
       -net none -chardev stdio,id=con,mux=on -serial chardev:con \
       -mon chardev=con,mode=readline -kernel ./nuttx

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



The provided PR description is **insufficient** based on the NuttX requirements. 

Here's why and how to improve it:

**Missing Information**

* **Summary:**
    * **Why is this change necessary?**  What specific problems does using percpu storage for task information solve?  Is there a performance bottleneck this addresses? Is it a prerequisite for another feature?
    * **What functional part of the code is being changed?** Be more specific than "arm64". Are you modifying the scheduler?  A specific architecture port?
* **Impact:**
    * **Is this a new feature or changing an existing one?**  This is important for understanding the scope.
    * **Impact on user:** Even if you believe the impact on the user is minimal, state that explicitly. If there's no impact on the user API, say so.
    * **Impact on build:**  Will any new configuration options be added?
    * **You only mention "arm64".**  Will this impact other architectures? If so, how? If not, state that this is arm64 specific.
    * **Impact on documentation:**  New features or significant changes require documentation updates.  Indicate if you are providing documentation updates as part of this PR.
    * **Security, compatibility:** Even if there is no impact, state this explicitly.
* **Testing:**
    * **Insufficient build host details:** Specify the operating system (e.g., Ubuntu 20.04), CPU architecture (x86_64), and compiler version used for testing.
    * **"Running with qemu" is not enough:**
        * Provide the full QEMU command line used.
        * **Crucially, you need to show HOW you tested this.** What specific commands did you run in the NuttX shell (`nsh`) to validate that `up_this_task` and `up_update_task` work as expected?
    * **Show both "before" and "after" logs.**  What is different in the output that demonstrates the change is working?

**Recommendations**

1. **Provide concrete motivations for the change.**
2. **Be specific about the code being modified.**
3. **Explain the user impact even if it is minimal.**
4. **Detail build system changes.**
5. **Clarify architecture specificity (is it only arm64?).**
6. **Describe any new or modified documentation.**
7. **Provide comprehensive test cases with expected output.**  Demonstrate how you validated that the percpu storage is being used correctly.

By addressing these points, your PR will be much clearer and easier for the NuttX maintainers to review.

File: sched/irq/irq_csection.c [hunk 1]
```diff
@@ -290,6 +290,9 @@ irqstate_t enter_critical_section(void)
                   DEBUGVERIFY(up_cpu_paused(cpu));
                   paused = true;
 
+                  /* After resume current_task may change */
+
+                  up_update_task(current_task(cpu));
```
Comment (pussuw) [inline sched/irq/irq_csection.c:295]: After resume, do we already know which task is going to be scheduled ?

File: include/nuttx/arch.h [hunk 2]
```diff
@@ -930,6 +930,43 @@ int up_copy_section(FAR void *dest, FAR const void *src, size_t n);
...
+ *
+ * Input Parameters:
+ *   current tcb
+ *
+ * Returned Value:
+ *   current tcb
+ *
+ ****************************************************************************/
+
+#ifndef up_update_task
```
Comment (pussuw) [inline include/nuttx/arch.h:966]: It seems this macro is implemented in the next patch, I suggest squashing them as they are the same feature.

File: sched/sched/sched.h [hunk 3]
```diff
@@ -379,7 +378,11 @@ void nxsched_sporadic_lowpriority(FAR struct tcb_s *tcb);
 void nxsched_suspend(FAR struct tcb_s *tcb);
 #endif
 
-#ifdef CONFIG_SMP
+#if defined(up_this_task)
+#  define this_task()            up_this_task()
+#elif !defined(CONFIG_SMP)
+#  define this_task()            ((FAR struct tcb_s *)g_readytorun.head)
```
Comment (pussuw) [inline sched/sched/sched.h:384]: Why not use (current_task(this_cpu())) like it was before ?

File: arch/arm64/include/arch.h [hunk 4]
```diff
@@ -52,6 +53,37 @@
 
 #endif /* CONFIG_ARCH_ADDRENV */
 
+/****************************************************************************
```
Comment (pussuw) [inline arch/arm64/include/arch.h:56]: I cannot see a reason in this patch why these were moved, arm64_arch is visible to modules in src/common. Why were these moved?

File: arch/arm64/include/irq.h [hunk 5]
```diff
@@ -397,6 +397,15 @@ static inline void up_irq_restore(irqstate_t flags)
 #  define up_cpu_index() (0)
 #endif
 
+/****************************************************************************
+ * Schedule acceleration macros
+ ****************************************************************************/
+
+#define up_current_regs()      (this_task()->xcp.regs)
+#define up_this_task()         ((struct tcb_s *)(read_sysreg(tpidr_el1) & ~1ul))
```
Comment (pussuw) [inline arch/arm64/include/irq.h:405]: What is the meaning of this magical lsbit check ?

File: arch/arm64/src/common/arm64_doirq.c [hunk 6]
```diff
@@ -61,13 +61,12 @@ uint64_t *arm64_doirq(int irq, uint64_t * regs)
...
 
-  DEBUGASSERT(up_current_regs() == NULL);
+  DEBUGASSERT(!up_interrupt_context());
 
-  /* Current regs non-zero indicates that we are processing an interrupt;
-   * current_regs is also used to manage interrupt level context switches.
-   */
+  /* Set irq flag */
+
+  write_sysreg((uintptr_t)tcb | 1, tpidr_el1);
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_doirq.c:68]: What is the significance of setting the lsbit here ?

File: arch/arm64/src/common/arm64_cpustart.c [hunk 7]
```diff
@@ -226,6 +222,12 @@ int up_cpu_start(int cpu)
 
 void arm64_boot_secondary_c_routine(void)
 {
+  struct tcb_s *tcb = current_task(this_cpu());
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_cpustart.c:225]: I think most places use this_task(), does it not work here ?

File: include/nuttx/arch.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/nuttx/arch.h:966]: each arch need provides the different implementation and mayn't provide an implementation at all, so it's better that the first patch only introduces the hook point.

File: arch/arm64/src/common/arm64_doirq.c (hunk 6 above)
Comment (xiaoxiang781216) [reply arch/arm64/src/common/arm64_doirq.c:68]: it's used to indicate the code is running inside irq context.

File: arch/arm64/include/irq.h (hunk 5 above)
Comment (xiaoxiang781216) [reply arch/arm64/include/irq.h:405]: whether we are running inside irq handler.

File: include/nuttx/arch.h (hunk 2 above)
Comment (pussuw) [reply include/nuttx/arch.h:966]: Ok, makes sense

File: arch/arm64/include/irq.h (hunk 5 above)
Comment (pussuw) [reply arch/arm64/include/irq.h:405]: Ok, I suggest giving a descriptive name to the magic number

File: sched/sched/sched.h (hunk 3 above)
Comment (hujun260) [reply sched/sched/sched.h:384]: #define this_task() (current_task(this_cpu())) 
This macro can only be used in a single-core environment.

File: sched/irq/irq_csection.c (hunk 1 above)
Comment (hujun260) [reply sched/irq/irq_csection.c:295]: current_task(this_cpu()) is the running task going to be scheduled

File: arch/arm64/include/arch.h (hunk 4 above)
Comment (hujun260) [reply arch/arm64/include/arch.h:56]: arch/arm64/src/common/arm64_arch.h ，cannot be seen in the public code.
eg.
irq_csection.c ...

File: arch/arm64/src/common/arm64_cpustart.c (hunk 7 above)
Comment (hujun260) [reply arch/arm64/src/common/arm64_cpustart.c:225]: it not work here, 
 Because we need to complete the initialization percpu before using percpu, and this is where the initialization takes place.

File: arch/arm64/include/arch.h (hunk 4 above)
Comment (pussuw) [reply arch/arm64/include/arch.h:56]: Ok

File: arch/arm64/include/irq.h [hunk 8]
```diff
@@ -387,31 +387,17 @@ static inline void up_irq_restore(irqstate_t flags)
...
-  __asm__ volatile ("mrs %0, " "tpidr_el1" : "=r" (regs));
-  return regs;
-}
-
-noinstrument_function
-static inline_function void up_set_current_regs(uint64_t *regs)
-{
-  __asm__ volatile ("msr " "tpidr_el1" ", %0" : : "r" (regs));
-}
+#define up_current_regs()      (this_task()->xcp.regs)
```
Comment (xiaoxiang781216) [inline arch/arm64/include/irq.h:397]: let's remove up_current_regs from common code

Comment (GUIDINGLI) [reply arch/arm64/include/irq.h:392]: Next PR

**Comment by GUIDINGLI** [comment]
LGTM
